### PR TITLE
[Compression] Quantization: add module fusion

### PR DESCRIPTION
**Pull request by Bonytu**
### Description ###
module fusion for quant
#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

File: nni/contrib/compression/utils/fused_utils.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+def get_module(model: torch.nn.Module, module_name: str):
+    module_name_list = module_name.strip().split('.')
+    cur_module = model
+    for s in module_name_list:
+        cur_module = getattr(cur_module, s, None)
+        if cur_module is None:
+            raise ValueError(f"can\'t find {module_name} in the model")
+
+    return cur_module
```
Comment (J-shang) [inline nni/contrib/compression/utils/fused_utils.py:23]: here can use `get_nested_attr`

File: nni/contrib/compression/example/test_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    'granularity': 'default',
+}]
+
+config_list_2 = [{
+    'op_names': ['conv1'],
+    'target_names': ['weight', '_input_', "_output_", "bias"],
+    'quant_dtype': 'int2',
+    'quant_scheme': 'affine',
+    'granularity': 'default',
+    'fuse_names': ["conv1", "batchnorm1"]
```
Comment (J-shang) [inline nni/contrib/compression/example/test_wrapper.py:76]: a little confused, maybe [("conv1", "batchnorm1")]?

File: nni/contrib/compression/base/wrapper.py [hunk 3]
```diff
@@ -75,6 +87,33 @@ def __init__(self, module: torch.nn.Module, module_name: str, config: Dict[str,
...
         self._track_funcs: List[Callable[[ModuleWrapper, str, Tensor], None]] = [track_target_shape]
 
+        # register for module fusion
+        self.register_fusion_info(fused_modules)
+
+    def register_fusion_info(self, fused_modules: List[torch.nn.Module] | None = None):
+        self.fused_modules = fused_modules if fused_modules is not None else []
+        if len(self.fused_modules) > 0:
+            self.is_bias = check_bias(self.module) # used for fold_bn
+            self.register_bias()
```
Comment (J-shang) [inline nni/contrib/compression/base/wrapper.py:97]: call this func in `wrap()`? we don't wrap the module at this time

File: nni/contrib/compression/base/wrapper.py [hunk 4]
```diff
@@ -125,6 +164,18 @@ def unwrap(self):
...
+            delattr(self.module, 'bias')
+            original_bias = self.module.original_bias
+            if isinstance(original_bias, nn.parameter.Parameter):
+                self.module.register_parameter('bias', torch.nn.Parameter(original_bias.detach().clone()))
+            elif isinstance(original_bias, torch.Tensor):
+                self.module.register_buffer('bias', original_bias.detach().clone())
+            delattr(self.module, 'original_bias')
+        if len(self.fused_modules) > 0 and not self.is_bias and check_bias(self.module):
+            delattr(self.module, 'bias')
+            self.module.register_parameter('bias', None)
```
Comment (J-shang) [inline nni/contrib/compression/base/wrapper.py:178]: what if the original module don't have a None bias?

File: nni/contrib/compression/base/wrapper.py (hunk 3 above)
Comment (Bonytu) [reply nni/contrib/compression/base/wrapper.py:97]: okok

File: nni/contrib/compression/base/wrapper.py (hunk 4 above)
Comment (J-shang) [reply nni/contrib/compression/base/wrapper.py:178]: I mean that some module like `nn.Softmax` do not have a parameter bias?
maybe `check_bias` should have three states: 'Tensor', 'None', 'non-exist'?

File: nni/contrib/compression/base/wrapper.py (hunk 3 above)
Comment (Bonytu) [reply nni/contrib/compression/base/wrapper.py:97]: Consider a case, if a module has wrapped in the pruning compressor, then it can not be wrapped for module fusion.
WARNING: Wrapper of module is wrapped, no need to wrap again.

File: nni/contrib/compression/example/test_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+module_wrappers_1, target_spaces_1 = register_wrappers(model, config_list_1, "quantization")
+print(f"target_space_1={target_spaces_1}\n")
+for module_name, wrapper in module_wrappers_1.items():
+    print(f"module_name={module_name}\tconfig={wrapper.config}\twrapper={wrapper}\n")
+
+module_wrappers_2, target_spaces_2 = register_wrappers(model, config_list_2, "quantization", module_wrappers_1)
+print(f"target_space_2={target_spaces_2}\n")
+for module_name, wrapper in module_wrappers_2.items():
+    print(f"module_name={module_name}\tconfig={wrapper.config}\twrapper={wrapper}\n")
+print(f"target_space_2={target_spaces_2}\n")
```
[the remark this answers is not in the document]
Comment (Bonytu) [reply nni/contrib/compression/example/test_wrapper.py]: done

File: nni/contrib/compression/base/config.py [hunk 6]
```diff
@@ -177,7 +180,10 @@ def select_modules_by_config(model: torch.nn.Module, config: Dict[str, Any]) ->
     for op_type in exclude_op_types:
         op_names.difference_update(type2names.get(op_type, set()))
 
-    return {module_name: name2module[module_name] for module_name in op_names}, config
+    if len(fuse_names) > 0 and len(op_names) > 1:
```
[the remark this answers is not in the document]
Comment (Bonytu) [reply nni/contrib/compression/base/config.py]: fixed

File: nni/contrib/compression/utils/fused_utils.py (hunk 1 above)
Comment (Bonytu) [reply nni/contrib/compression/utils/fused_utils.py]: okok

File: nni/contrib/compression/example/test_wrapper.py (hunk 2 above)
Comment (Bonytu) [reply nni/contrib/compression/example/test_wrapper.py]: done

File: nni/contrib/compression/base/wrapper.py [hunk 7]
```diff
@@ -488,7 +488,10 @@ def create_module_wrapper(model:nn.Module, module: nn.Module, module_name: str,
         target_spaces = wrapper.extend_target_spaces(config, mode)
         wrapper.config = update_config(wrapper.config, {mode: config})
         if len(fused_modules_pair) > 0:
-            wrapper.register_fusion_info(fused_modules)
+            assert hasattr(wrapper.module, '_nni_wrapper'), \
+                f'wrapper {wrapper.name} is not wrapped, please wrap it before register new wrapper'
+            wrapper.fused_modules = fused_modules
```
Comment (J-shang) [inline nni/contrib/compression/base/wrapper.py:493]: what if wrapper already has none empty `fused_modules`?

**Review by J-shang — APPROVED** [review]

**Review by QuanluZhang — APPROVED** [review]

File: nni/contrib/compression/base/config.py [hunk 8]
```diff
@@ -177,7 +180,7 @@ def select_modules_by_config(model: torch.nn.Module, config: Dict[str, Any]) ->
     for op_type in exclude_op_types:
         op_names.difference_update(type2names.get(op_type, set()))
 
-    return {module_name: name2module[module_name] for module_name in op_names}, config
+    return {module_name: name2module[module_name] for module_name in op_names}, config, fuse_names
```
Comment (J-shang) [inline nni/contrib/compression/base/config.py:183]: please search `select_modules_by_config` and check this retrun values change is legal.
